### ISSUE TITLE
Check if all functions are used in the codebase

### ISSUE THREAD

**Issue by PooyaHekmati** [RuminantFarmSystems/RuFaS#1105]
Let's do some housekeeping: are all functions used? Can we identify the stale functions?
You may want to write a little script that sweeps the codebase, finds the function definitions and stores them in dictionary. Then does another sweep and count how many times each function has been used. Be aware of the strings though, it is possible that the name of a function is used in an `info_map` or error message or warning. So simply checking if there is a match won't work. In scenarios like this, regular expressions are your best friends. Also, chatGPT suggested the following:

> To check if all functions are used in a Python codebase, you can use various approaches depending on the complexity and size of the codebase. One common approach is to use static code analysis tools. These tools can help identify unused functions, classes, and other elements in the code. Here are a few steps and tools you might consider:
> 
> 1. **Pyflakes or Flake8**: These tools are primarily used for checking Python code for errors and style issues, but they can also identify unused imports and variables, which can be a good starting point.
> 
> 2. **Vulture**: This is a more specialized tool for finding dead code (unused functions, classes, variables, etc.) in Python projects. Vulture is more aggressive in finding unused code compared to Pyflakes or Flake8.
> 
> 3. **Custom Scripting**: For a more tailored solution, especially in complex projects with specific requirements, you can write a custom script using Python's `ast` (Abstract Syntax Tree) module. This approach allows you to analyze the codebase and identify unused functions by traversing the AST nodes.
> 
> 4. **Integration with Your IDE**: Some Integrated Development Environments (IDEs) like PyCharm have built-in or plugin-based tools to detect unused code.
> 
> 5. **Test Coverage Tools**: Although not directly intended for finding unused functions, tools like `coverage.py` can be helpful. By running your test suite with coverage analysis, you can identify parts of your code that are never executed, which may include unused functions.
> 
> 6. **Manual Code Review**: In addition to automated tools, a thorough manual code review can sometimes identify unused or redundant code that automated tools might miss.
> 
> Remember that these tools might have limitations and might not always accurately identify all unused code, especially in dynamic languages like Python where certain usages can be obscured through reflection, dynamic imports, or other metaprogramming techniques. Always review the findings of these tools and test your application thoroughly after removing any code.

**Comment by Isgotkowitz** [comment]
A related issue, that may or may not be addressed alongside efforts to address this issue, is checking if all variables/attributes are being used in the codebase.

**Comment by Isgotkowitz** [comment]
@JoeWaddell has worked on this issue before. He will have some insights into how to approach this.

**Comment by ew3361zh** [comment]
Unused functions (from Vulture package scan):

RUFAS/routines/animal/clustering_pen_grouping.py:38: unused function 'grouping' (60% confidence)
RUFAS/routines/animal/manure/general_manure.py:119: unused function 'scalar_mult_animal_manure_excretions' (60% confidence)

More analyses coming.

**Comment by ew3361zh** [comment]
Ran a script using the [Python ast module](https://docs.python.org/3/library/ast.html#function-and-class-definitions) to collect all the function names by python file. I tried to exclude all the packages/plug-ins that RuFaS uses as best I could. Here is the list of all the RuFaS functions as of running this script Feb 7, 2024:

[All RuFaS Functions](https://docs.google.com/document/d/1dtq7zJGIonZ6tjbzvVDYRkhiUVtm4eaSjfE-eTP9BmY/edit)

Ran a separate script incorporating the [Python cProfile module](https://docs.python.org/3/library/profile.html#module-cProfile) to collect all the functions that were called during a simulation. I again tried to exclude all the plug-ins/packages that were not RuFaS-written. From that, I compared the list of all RuFaS functions to the list of called functions and came up with the list of functions that were not called at runtime:

[RuFaS Functions Uncalled at Runtime](https://docs.google.com/document/d/1NroaHk1C9DDKxRtudM6MKKgyX31tC2PXiYClEirpt_Y/edit)

Notes:
- A number of the functions that were uncalled are specific-scenario dependent. They may not have been called during the simulation I ran using the default scenario metadata settings but they may still be active in other scenarios.
- Comparing the two lists proved to be one of the more difficult tasks as the outputs from these packages were not designed for comparison. Some of this had to be done by hand. I have a pretty high level of confidence in the accuracy of the uncalled functions list as determined by the differences between these two lists.
- Further analysis is needed for any next steps. I will create an issue to analyze and move forward with next steps on these results.
